src/TableGatewayAbstractFactory.php

```diff
@@ -60,57 +92,120 @@ public function createServiceWithName(ServiceLocatorInterface $services, $name,
+    public function createServiceWithName(ServiceLocatorInterface $container, $name, $requestedName)
     {
-        if (!isset($config['table_name'])) {
+        return $this($container, $requestedName);
+    }
+
+    /**
+     * Is the configuration valid?
+     *
+     * @param array $config
+     * @param ContainerInterface $container
+     * @return bool
+     */
+    protected function isValidConfig(array $config, ContainerInterface $container)
+    {
+        if (! isset($config['table_name'])) {
             return false;
         }
 
         if (isset($config['adapter_name'])
-            && $services->has($config['adapter_name'])
+            && $container->has($config['adapter_name'])
         ) {
             return true;
         }
 
-        if (!isset($config['adapter_name'])
-            && $services->has('Zend\Db\Adapter\Adapter')
+        if (! isset($config['adapter_name'])
+            && ($container->has(AdapterInterface::class) || $container->has(Adapter::class))
         ) {
             return true;
         }
 
         return false;
     }
 
-    protected function getAdapterFromConfig(array $config, ServiceLocatorInterface $services)
+    /**
+     * Retrieve a zend-db adapter via provided configuration.
+     *
+     * If the configuration defines an `adapter_name` and a matching service
+     * is discovered, that will be returned.
+     *
+     * If the Adapter service is present, that will be returned (zend-mvc v2).
+     *
+     * Otherwise, the AdapterInterface service is returned.
+     *
+     * @param array $config
+     * @param ContainerInterface $container
+     */
+    protected function getAdapterFromConfig(array $config, ContainerInterface $container)
     {
         if (isset($config['adapter_name'])
-            && $services->has($config['adapter_name'])
+            && $container->has($config['adapter_name'])
         ) {
-            return $services->get($config['adapter_name']);
+            return $container->get($config['adapter_name']);
+        }
+
+        if ($container->has(Adapter::class)) {
+            // v2 usage
+            return $container->get(Adapter::class);
         }
 
-        return $services->get('Zend\Db\Adapter\Adapter');
+        // v3 usage
+        return $container->get(AdapterInterface::class);
     }
 
-    protected function getHydratorFromConfig(array $config, ServiceLocatorInterface $services)
+    /**
+     * Retrieve the configured hydrator.
+     *
+     * If configuration defines a `hydrator_name`, that service will be
+     * retrieved from the HydratorManager; otherwise ArraySerializable
+     * will be retrieved.
+     *
+     * @param array $config
+     * @param ContainerInterface $container
+     * @return \Zend\Hydrator\HydratorInterface
+     */
+    protected function getHydratorFromConfig(array $config, ContainerInterface $container)
     {
         $hydratorName = isset($config['hydrator_name']) ? $config['hydrator_name'] : 'ArraySerializable';
```

Comment:
No, because the class could be either `Zend\Stdlib\Hydrator\ArraySerializable` or `Zend\Hydrator\ArraySerializable`.